source/adaptive_load/metrics_evaluator_impl.cc

```diff
@@ -0,0 +1,114 @@
+using ::nighthawk::adaptive_load::ThresholdSpec;
+
+} // namespace
+
+absl::StatusOr<nighthawk::adaptive_load::MetricEvaluation>
+MetricsEvaluatorImpl::EvaluateMetric(const MetricSpec& metric_spec, MetricsPlugin& metrics_plugin,
+                                     const ThresholdSpec* threshold_spec) const {
+  nighthawk::adaptive_load::MetricEvaluation evaluation;
+  evaluation.set_metric_id(
+      absl::StrCat(metric_spec.metrics_plugin_name(), "/", metric_spec.metric_name()));
+  const absl::StatusOr<double> metric_value_or =
+      metrics_plugin.GetMetricByName(metric_spec.metric_name());
+  if (!metric_value_or.ok()) {
+    return absl::Status(static_cast<absl::StatusCode>(metric_value_or.status().code()),
+                        absl::StrCat("Error calling MetricsPlugin '",
+                                     metric_spec.metrics_plugin_name(), ": ",
+                                     metric_value_or.status().message()));
+  }
+  const double metric_value = metric_value_or.value();
+  evaluation.set_metric_value(metric_value);
+  if (threshold_spec == nullptr) {
+    // Informational metric.
+    evaluation.set_weight(0.0);
+  } else {
+    evaluation.set_weight(threshold_spec->weight().value());
+    absl::StatusOr<ScoringFunctionPtr> scoring_function_or =
+        LoadScoringFunctionPlugin(threshold_spec->scoring_function());
+    RELEASE_ASSERT(scoring_function_or.ok(),
+                   absl::StrCat("ScoringFunction plugin loading error should have been caught "
+                                "during input validation: ",
+                                scoring_function_or.status().message()));
+    ScoringFunctionPtr scoring_function = std::move(scoring_function_or.value());
+    evaluation.set_threshold_score(scoring_function->EvaluateMetric(metric_value));
+  }
+  return evaluation;
+}
+
+const std::vector<std::pair<const MetricSpec*, const ThresholdSpec*>>
+MetricsEvaluatorImpl::ExtractMetricSpecs(
+    const nighthawk::adaptive_load::AdaptiveLoadSessionSpec& spec) const {
+  std::vector<std::pair<const MetricSpec*, const ThresholdSpec*>> spec_threshold_pairs;
+  for (const MetricSpecWithThreshold& metric_threshold : spec.metric_thresholds()) {
+    spec_threshold_pairs.emplace_back(&metric_threshold.metric_spec(),
+                                      &metric_threshold.threshold_spec());
+  }
+  for (const MetricSpec& metric_spec : spec.informational_metric_specs()) {
+    spec_threshold_pairs.emplace_back(&metric_spec, nullptr);
+  }
+  return spec_threshold_pairs;
+}
+
+absl::StatusOr<nighthawk::adaptive_load::BenchmarkResult>
+MetricsEvaluatorImpl::AnalyzeNighthawkBenchmark(
+    const nighthawk::client::ExecutionResponse& nighthawk_response,
+    const nighthawk::adaptive_load::AdaptiveLoadSessionSpec& spec,
+    const absl::flat_hash_map<std::string, MetricsPluginPtr>& name_to_custom_metrics_plugin_map)
+    const {
+  if (nighthawk_response.error_detail().code() != static_cast<int>(absl::StatusCode::kOk)) {
```

Comment:
I'm sure this is tested so I'm probably wrong, but isn't error_detail().code() not guaranteed to exist here?  What happens if it doesn't?

If there's any concern of that creating weird behavior, we could wrap it with a has_error_detail / has_code

Reply:
We will get the default instance of the `error_detail` field if it's not set, and that should have `code` 0.